ledger-service/http-json/src/main/scala/com/digitalasset/http/endpoints/MeteringReportEndpoint.scala

```diff
@@ -0,0 +1,199 @@
+    override def write(obj: LocalDate): JsValue = JsString(obj.toString)
+    override def read(json: JsValue): LocalDate =
+      LocalDate.parse(json.asInstanceOf[JsString].value)
+  }
+
+  implicit val ParticipantIdFormat: RootJsonFormat[ParticipantId] =
+    new RootJsonFormat[ParticipantId] {
+      override def write(obj: ParticipantId): JsValue = JsString(obj)
+      override def read(json: JsValue): ParticipantId =
+        ParticipantId.assertFromString(json.asInstanceOf[JsString].value)
+    }
+
+  implicit val ApplicationIdFormat: RootJsonFormat[ApplicationId] =
+    new RootJsonFormat[ApplicationId] {
+      override def write(obj: ApplicationId): JsValue = JsString(obj)
+      override def read(json: JsValue): ApplicationId =
+        ApplicationId.assertFromString(json.asInstanceOf[JsString].value)
+    }
+
+  import spray.json.DefaultJsonProtocol._
+
+  implicit val MeteringReportDateRequestFormat: RootJsonFormat[MeteringReportDateRequest] =
+    jsonFormat3(MeteringReportDateRequest.apply)
+
+  implicit val MeteringReportRequestFormat: RootJsonFormat[MeteringReportRequest] =
+    jsonFormat3(MeteringReportRequest.apply)
+
+  implicit val ApplicationMeteringReportFormat: RootJsonFormat[ApplicationMeteringReport] =
+    jsonFormat2(ApplicationMeteringReport.apply)
+
+  implicit val MeteringReportFormat: RootJsonFormat[MeteringReport] =
+    jsonFormat4(MeteringReport.apply)
+
+  private[endpoints] def toTimestamp(pbTimestamp: protobuf.timestamp.Timestamp): Timestamp = {
+    Timestamp.assertFromInstant(
+      Instant.ofEpochSecond(pbTimestamp.seconds, pbTimestamp.nanos.toLong)
+    )
+  }
+
+  private val startOfDay = LocalTime.of(0, 0, 0)
+  private[endpoints] def toTimestamp(ts: LocalDate): Timestamp = {
+    Timestamp.assertFromInstant(Instant.ofEpochSecond(ts.toEpochSecond(startOfDay, ZoneOffset.UTC)))
+  }
+
+  private[endpoints] def toPbTimestamp(ts: Timestamp): protobuf.timestamp.Timestamp = {
+    val instant = ts.toInstant
+    protobuf.timestamp.Timestamp(instant.getEpochSecond, instant.getNano)
+  }
+
+  private[endpoints] def toPbRequest(
+      request: MeteringReportRequest
+  ): metering_report_service.GetMeteringReportRequest = {
+    import request._
+    metering_report_service.GetMeteringReportRequest(
+      from = Some(toPbTimestamp(from)),
+      to = to.map(toPbTimestamp),
+      applicationId = application.fold(
+        metering_report_service.GetMeteringReportRequest.defaultInstance.applicationId
+      )(identity),
+    )
+  }
+
+  // With Protobuf optional string is represented as an empty strings
+  private def pbOption(str: String): Option[String] = Option(str).filter(_.nonEmpty)
+
+  private def mustHave[T](option: Option[T], field: String): Either[String, T] = option match {
+    case Some(t) => Right(t)
+    case None => Left(s"GetMeteringReportResponse missing field, expected $field")
+  }
+
+  type StringOr[T] = Either[String, T]
```

Comment:
In general I discourage type aliases: this removes just a few characters and in turn it forces you to refer to the type alias definition if you don't know what a `StringOr` is. Conversely, I assume `Either` is supposed to be common knowledge, since it's part of the standard library.

Reply:
I added this to help out IDEA that often struggles to understand what is going on when classes have multiple type parameters, for example when traversing)

As adding clarity for IDEA in this case causes clouds the human reading of the code I've removed StringOr and the associated explicit type parameters use.

Reply:
I agree with @stefanobaghino-da here.

Reply:
@S11001001 is not a fan either, it has now been removed.